### PR TITLE
-[HTTPServer start:] will no longer fail if it can't bind an IPv6 socket

### PR DESCRIPTION
**Pull request by ksuther**
We had a user with IPv6 completely disabled (on Mac), which was causing `-[HTTPServer start:]` to fail.
Since only the IPv4 loopback address is used, I've made `start:` succeed even if it can't bind an IPv6 socket.

I also added the option of using the IPv6 loopback address (`::1`) in the future if iPv6 should be preferred over IPv4.

### CODE REVIEW

File: Source/macOS/LoopbackHTTPServer/OIDLoopbackHTTPServer.m [hunk 1]
```diff
@@ -585,4 +583,14 @@ - (BOOL)stop {
     return YES;
 }
 
+- (BOOL)hasIPv4Socket
```
Comment (StevenEWright) [inline Source/macOS/LoopbackHTTPServer/OIDLoopbackHTTPServer.m]: For consistency:

`- (BOOL)hasIPv4Socket {`

File: Source/macOS/LoopbackHTTPServer/OIDLoopbackHTTPServer.m [hunk 2]
```diff
@@ -585,4 +583,14 @@ - (BOOL)stop {
...
+{
+    return ipv4socket != nil;
+}
+
+- (BOOL)hasIPv6Socket
```
Comment (StevenEWright) [inline Source/macOS/LoopbackHTTPServer/OIDLoopbackHTTPServer.m]: For consistency:

`- (BOOL)hasIPv6Socket`

File: Source/macOS/LoopbackHTTPServer/OIDLoopbackHTTPServer.h [hunk 3]
```diff
@@ -72,6 +72,9 @@ typedef enum {
 - (BOOL)start:(NSError **)error;
 - (BOOL)stop;
 
+- (BOOL)hasIPv4Socket;
```
Comment (StevenEWright) [inline Source/macOS/LoopbackHTTPServer/OIDLoopbackHTTPServer.h]: Since `start:` always returns `NO` if it can't create an IPv4 socket, why is this necessary?

Comment (StevenEWright) [inline Source/macOS/LoopbackHTTPServer/OIDLoopbackHTTPServer.h]: Why not something like:

`@property(nonatomic, readonly) NSURL *preferredLoopbackURL;`

File: Source/macOS/OIDRedirectHTTPHandler.m [hunk 4]
```diff
@@ -76,10 +76,18 @@ - (NSURL *)startHTTPListener:(NSError **)returnError {
       *returnError = error;
     }
     return nil;
-  } else {
+  } else if ([_httpServ hasIPv4Socket]) {
```
Comment (StevenEWright) [inline Source/macOS/OIDRedirectHTTPHandler.m]: The order of these clauses would cause the library to always use IPv4, since, with your changes, the library always must have an IPv4 socket, but only may have an IPv6 socket, this first test will always pass, leaving no chance for the second test.

Comment (ksuther) [inline Source/macOS/OIDRedirectHTTPHandler.m]: This was intentional, as I didn't want to change the behavior of the library dramatically without any discussion. Handling a potential IPv6-only case is easy enough to do, I'll add it even though it's unlikely to be hit any time soon.

Comment (StevenEWright) [inline Source/macOS/OIDRedirectHTTPHandler.m]: That would be very(!) much appreciated.

I do know Apple is concerned with IPv6-only networks ([see here](https://developer.apple.com/support/ipv6/)), so I think the feature may be more valuable than you're giving it credit for. IMHO, the lack of this support is a bug with the existing code.

**Comment by ksuther** [comment]
`start:` will now succeed if IPv6 is available but IPv4 isn't. The callback URL still prefers IPv4 over IPv6 if both are available, but that could be changed by swapping the order of the check in `OIDRedirectHTTPHandler`.

File: Source/macOS/LoopbackHTTPServer/OIDLoopbackHTTPServer.h (hunk 3 above)
Comment (ksuther) [inline Source/macOS/LoopbackHTTPServer/OIDLoopbackHTTPServer.h]: I don't know that there's much gain to splitting out part of `startHTTPListener:` into a separate property. The method is only a few lines as is.

**Comment by StevenEWright** [comment]
This PR LGTM.

@ksuther Thank you! Were you able to test it, or should I patch it in and try it?

File: Source/macOS/LoopbackHTTPServer/OIDLoopbackHTTPServer.h (hunk 3 above)
Comment (StevenEWright) [inline Source/macOS/LoopbackHTTPServer/OIDLoopbackHTTPServer.h]: This comment was more to do with making life easy for consumers of this class - since the only reason consumers care about these values is in the context of determining which loopback address to use.

**Comment by ksuther** [comment]
I tested it by making the IPv4 socket fail, but I didn't actually turn off IPv4 systemwide.

**Comment by ksuther** [comment]
Does this require anything else from me?

**Comment by WilliamDenniss** [comment]
Do you want to squash the commits into one? I can do this for you but then I'm added as the "committer", so thought I'd give you first option.

**Comment by ksuther** [comment]
I squashed them, I hope that's what you were expecting.

**Comment by WilliamDenniss** [comment]
Perfect, thanks.
